### DISCUSSION TITLE
When can we see PhotoPrism as an app in the NextCloud app store?

### DISCUSSION THREAD

**Discussion by mvozzo** [photoprism/photoprism#2259] [Ideas]
It would be awesome to see PhotoPrism as a NextCloud App under the Multimedia section. 

I think PhotoPrism would quickly become the Number 1 photo & video management solution as nothing currently comes close.  Check out the existing Multimedia apps available on NextCloud here - https://apps.nextcloud.com/categories/multimedia

![_nextcloud-multimedia-apps](https://user-images.githubusercontent.com/10014870/163694759-cdaa6619-5e1c-45da-9218-768007fd78de.jpg)

Please consider this opportunity as I feel your software could benefit many thousands upon thousands of people. There is a movement away from using public cloud storage solutions that can be censored and spied upon to self-hosted storage solutions. I feel NextCloud are leader in this space and have a growing ecosystem and user base.  

I feel that PhotoPrism has a small window of opportunity to take the 1st mover advantage position within the Nextcloud ecosystem as their app library continues to grow in the future. 

NextCloud has already figured out the hard stuff of user uploading photos and has it's own mobile app on Apple and Android that facilities easy Camera Phone upload and sync.  So, all PhotoPrism devs need to figure out is how to hook into the photo and video file structure in NextCloud and do what PhotoPrism does best "... an AI-powered app for browsing, organizing & sharing your photo collection. It makes use of the latest technologies to tag and find pictures automatically without getting in your way."

**Comment by graciousgrey** [comment]
PhotoPrism can be used with Nextcloud already. Many of our users do it by configuring Nextcloud's photos folder as PhotoPrism's originals directory :)

**Reply by mvozzo** [reply]
Yes, I know this to be the case and it supports the point I am trying to make. This discussion is more about making PhotoPrism more discoverable within the NextCloud ecosystem. 

If PhotoPrism was available as a "1-click install" from the NextCloud App Library, I think this will expand your user base exponentially.   

As you rightly mention, PhotoPrism already works with NextCloud, so porting PhotoPrism to NextCloud's app library should be easy, right?  

Making your software super easy to install from within NextCloud will give PhotoPrism a 1st mover advantage as it will be the only and the best photo & video management application within the NextCloud ecosystem.  

Have you seen what alternatives there are in the "Multimedia" section ( https://apps.nextcloud.com/categories/multimedia ) in NextCloud's app library?  Let me make it easy for you... None.  Hence, I think this will be a huge opportunity for PhotoPrism to expand it's user base and brand.

**Reply by lastzero** [reply]
From a maintainer's perspective, it's odd that Nextcloud is working in the exact same area without having approached us (we tried early on by tweeting about them and including links to Nextcloud in our docs, with no reaction). The Nextcloud Recognize app is even partially built on our work:

- https://apps.nextcloud.com/apps/recognize
- https://github.com/marcelklehr/recognize/blob/ef0ed570e4e4926bfe8672e7c4b4f34de20ad044/src/rules.yml

If the company behind Nextcloud sees us as competition and prefers to develop their own solution instead of working with us, I see no reason to invest time in an app for their store. It may give us more visibility, but development time is our scarcest resource and they might as well reject a Nextcloud PhotoPrism app.

**Reply by mvozzo** [reply]
Wow, I didn't even realise the Recognise app is based on work done by PhotoPrism. Thanks for sharing that information. 

I'm curious, has PhotoPrism submitted this form - https://apps.nextcloud.com/developer/apps/generate  in an attempt to start some dialogue with NextCloud?  

Or perhaps, has anyone on the PhotoPrism team reached out to the maintainer of the "NextCloud Photos" app - https://github.com/nextcloud/photos 
You never know, there could be an opportunity to collaborate/integrate PhotoPrism? 

I can't see why NextCloud would want to develop and maintain there own version of a PhotoPrism-like application they are not in the business of photo and video management, you guys are the experts and leaders in this space.  The NextCloud dev are providing the platform that apps sit on top of.  Perhaps you should consider reaching out to them again by submitting the form above and see where that leads.  

Is there anything I can possibly do to help start some dialogue between PhotoPrism and NextCloud?  
In full disclosure I have no affiliation with NextCloud. I'm just a new user of Nextcloud and not a very techncial one at that. Hence my selfish interest would be to have a 1-click install of PhotoPrism in NextCloud as your software is 10 million times better at photo and video management than the out-of-the-box NextCloud Photos app and I think having it readily available in the NextCloud App library will shine a light on your product and help grow your user base.

**Reply by graciousgrey** [reply]
Thank you very much for your kind feedback and offering your help. We will consider this at a later time, as we are extremely busy at the moment.

**Reply by marcelklehr** [reply]
Hey everyone, this discussion was brought to my attention recently and I thought I'd chime in as my app was mentioned.

> From a maintainer's perspective, it's odd that Nextcloud is working in the exact same area without having approached us (we tried early on by tweeting about them and including links to Nextcloud in our docs, with no reaction). The Nextcloud Recognize app is even partially built on our work

I should make clear that I am in no way affiliated with Nextcloud, and neither is the recognize app. Please don't take the existence of recognize as a sign that Nextcloud wants to compete with you, doesn't want to collaborate or will reject an app submission. I cannot speak for the company or the project.

My view on this topic in general, if I may, is that it's hard to integrate existing apps into nextcloud without some drawbacks, which is likely why nobody has done this for photoprism. Photoprism, as I understand it, is built as a separate app, that you log into, store your data in etc. I think photoprism excels at what it's intended to do, but I'm not sure a Nextcloud app that packages photoprism would excel as a Nextcloud app, because it would probably still be kind of separate, unless extensive efforts are undertaken to bridge the gaps. This is why I opted to go down a different route, which is to take inspiration from photoprism (and other apps) but build an app that is as deeply integrated into the nextcloud ecosystem as possible. (Which also has drawbacks, by the way, cause boy do I hate PHP...)

Peaceful greetings,
Marcel

**Reply by sunjam** [reply]
I'd like to chime in with some thoughts:
Very interesting conversation.  Worth mentioning that the company Nextcloud remains focused within their github repos afaik, so my guess is they know nothing about your discussion here...  the good news is they do offer [a community chat](https://cloud.nextcloud.com/call/xs25tz5y) along with their irc channel for developers.  This is an excellent way to contact them and increase awareness.  Hope it helps.

**Reply by FatespinnerX9** [reply]
> PhotoPrism can be used with Nextcloud already. Many of our users do it by configuring Nextcloud's photos folder as PhotoPrism's originals directory :)

I've been googling and searching to do exactly this. I'm running everything on Unraid are there instructions for the unwashed? 

My goal is to have all my users (all 2 of us) photos to end up in photoprism. Ideally we would also be able to look back at old stuff too.

**Reply by sunjam** [reply]
@FatespinnerX9 asking for support in this discussion topic is not appropriate.  [Please ask somewhere else](https://photoprism.app/contact).

**Reply by graciousgrey** [reply]
@FatespinnerX9 In general, you can mount multiple directories as subfolders inside originals. How to do that with docker-compose is described here: https://docs.photoprism.app/getting-started/docker-compose/#volumes. I'm not sure how to do this on unraid. Feel free to open a new discussion for this topic to find a community member who can help with this.

**Reply by 0rb1ta** [reply]
join, much needed app

**Reply by marcelklehr** [reply]
Here's an update from my side: I've since been hired by Nextcloud to work on Recognize full-time. My point still stands that photoprism is really cool and we love what you do, but it's hard to integrate it into Nextcloud as an app. Still, I'd be very happy to collaborate, share ideas and thoughts with you folks, if you'd be up for that. You can reach the photos developers (including me) over at https://cloud.nextcloud.com/call/m7tc7ndi :blue_heart:

**Reply by lastzero** [reply]
Congratulations! It would be great if we can collaborate in the future, maybe after our multi-user support is released so that PhotoPrism can be more easily integrated with other multi-user web apps like Nextcloud :cloud:

**Comment by sunjam** [comment]
You can always use the [external sites](https://apps.nextcloud.com/apps/external) app to access your photoprism instance from within Nextcloud as though it were "built in".  If you try this out, experiment with whether you'll need an iframe.  _Fyi, there is an issue where it is suggested to create a Nextcloud app as well_ #1271

Multi-user support and more authentication methods will be important steps for gradually introducing even more integration between the two projects.  :smiley: 

- https://github.com/photoprism/photoprism/issues/98
- https://github.com/photoprism/photoprism/issues/782
- https://github.com/photoprism/photoprism/issues/484

The two projects are fundamentally different, and I don't imagine Photoprism will be ported to PHP, but I am looking forward to refinements down the road.  Thanks for already including so much interoperability!

**Comment by maxiride** [comment]
I've mounted the Nextcloud Photo directory inside PhotoPrism originals, it works as expected however I have to manually fire a rescan for new photos to appear.

Is there a way to schedule re scans? Or a better mechanic to have new photos uploaded through Nextcloud appear in PhotoPrism?

**Reply by graciousgrey** [reply]
When you move files to the originals folder using webdav (many sync tools such as PhotoSync use webdav to sync files), indexing is triggered automatically after a safety delay.

https://docs.photoprism.app/user-guide/library/originals/#automatic-indexing

Otherwise you can set up a cron job or configure a scheduler.

https://docs.photoprism.app/getting-started/docker-compose/#introduction
https://dl.photoprism.app/docker/scheduler/

A file system watcher for automatic indexing independent of webdav is planned, but not implemented yet.

https://github.com/photoprism/photoprism/issues/2608

**Reply by maxiride** [reply]
Thanks for the timely reply!

**Comment by scottdotau** [comment]
While using WebDAV through for example, Nextcloud, works... It means file (photo and video) duplication from my understanding as it requires synchronisation and/or uploading to the 'service'. 

A nextcloud plugin would mean these files could be proxied to PhotoPrism, either replacing or integrating with the Photos and/or Memories apps. 

Working with the team of the memories app would probably benefit both products, as the memories app requires extra installation steps within nextcloud in order to produce essentially the same result as PhotoPrism, which is exponentially faster when dealing with media (prob coz not php)

**Reply by lastzero** [reply]
@scottdotau While we currently don't have resources to work on this, it seems like a great approach to offer/support alternative UIs like a version of the Memories app adapted to our API.

We could also extend our existing REST API for this purpose, as long as the changes can be implemented within a reasonable timeframe and do not compromise security, e.g. by adding new endpoints without sufficient authorization checks.

**Reply by scottdotau** [reply]
This seems extremely reasonable, thank you! If someone were to investigate writing a PhotoPrism app/integration for Nextcloud, you guys would support this as long as the above (resource and time constraints) weren't pushed? 

Mainly in regards to the API, as long as modifications are fairly generic/reusable within different applications?

**Reply by lastzero** [reply]
Exactly. Any alternative UI should of course also be usable without Nextcloud, so that users who would otherwise not need Nextcloud don't have to install it first.
